README
WHAT IS IT?
===========

AuVyon collects little helpers to work with media data.

See https://github.com/pyroscope/auvyon/wiki for more.


INSTALLATION
============

First, install basic requirements as root, e.g. on Debian and similar distros:

    sudo apt-get install \
        imagemagick ffmpeg flac \
        python-dev python-numpy python-scipy python-matplotlib \
        libsndfile1-dev libmagickwand-dev ttf-dejavu-extra \
        libatlas-dev liblapack-dev gfortran \
        python-virtualenv virtualenvwrapper git bzip2

If you can't find all of these packages for your distro, some features might not work.


Then, for installing a release version, just call

    easy_install auvyon

[NB: There currently *is* no release version]


Otherwise, to create a working directory from source, use

    cd; mkdir -p src; cd src
    git clone git://github.com/pyroscope/auvyon.git
    cd auvyon; export AUVYON_HOME="$PWD"
    . ./bootstrap.sh

This creates a virtualenv with all dependencies and necessary tools installed.


To test waveform creation, use
    $AUVYON_HOME/bin/python -m auvyon.imaging.waveforms $AUVYON_HOME/src/tests/data/Parrot_sketch.ogg 
    display $AUVYON_HOME/src/tests/data/waveforms.jpg
    # on a headless system, use "identify" instead of "display"

To test spectrogram creation, use
    $AUVYON_HOME/bin/python -m auvyon.imaging.spectrograms $AUVYON_HOME/src/tests/data/Parrot_sketch.ogg 
    display $AUVYON_HOME/src/tests/data/spectrogram.jpg
    # on a headless system, use "identify" instead of "display"



ACKNOWLEDGEMENTS
================

Repository hosting @ github.

Test data sources:
    http://en.wikipedia.org/wiki/File:Parrot_sketch.ogg



debian/changelog
auvyon (0.1) precise; urgency=low

  * first release

 -- pyroscope <pyroscope.project@gmail.com>  Sun, 03 May 2015 03:26:51 +0200


bootstrap.sh
# This script has to be sourced in a shell and is thus NOT executable.

# Helpers
have_pypkg() {
    ./bin/python -c "import $1" 2>/dev/null
}
need_pypkg() {
    have_pypkg "$1" && false || true
}

# Use or create virtualenv
if test -f ../bin/activate; then
    test -L bin -a -f bin/activate || ln -nfs ../bin .
elif test ! -f ./bin/activate; then
    python -c "import urllib2; open('virtualenv.py', 'w').write(\
        urllib2.urlopen('https://raw.github.com/pypa/virtualenv/master/virtualenv.py').read())"
    deactivate 2>/dev/null || true
    python virtualenv.py --system-site-packages . || return 1
fi

export DEBFULLNAME=pyroscope
export DEBEMAIL=pyroscope.project@gmail.com

grep DEBFULLNAME bin/activate >/dev/null || cat >>bin/activate <<EOF
export DEBFULLNAME=$DEBFULLNAME
export DEBEMAIL=$DEBEMAIL
EOF

. bin/activate || return 1

# Preconditions
ok=true
which montage >/dev/null || \
    { echo "ERROR: You MUST install ImageMagick (e.g. 'apt-get install imagemagick')!"; ok=false; }
which MagickWand-config >/dev/null || \
    { echo "ERROR: You MUST install MagicWand dev support (e.g. 'apt-get install libmagickwand-dev')!"; ok=false; }
which ffmpeg >/dev/null || \
    { echo "ERROR: You MUST install ffmpeg (e.g. 'apt-get install ffmpeg')!"; ok=false; }
which flac >/dev/null || \
    { echo "ERROR: You MUST install FLAC support (e.g. 'apt-get install flac')!"; ok=false; }
which git >/dev/null || \
    { echo "ERROR: You MUST install git (e.g. 'apt-get install git'; or possibly 'git-core')!"; ok=false; }
if which dpkg >/dev/null ; then
    dpkg -l ttf-dejavu-extra | grep ^ii >/dev/null || \
        { echo "ERROR: You MUST install ttf-dejavu-extra (e.g. 'apt-get install ttf-dejavu-extra')!"; ok=false; }
else
    echo "WARNING: Not a debian system, make sure you have 'ttf-dejavu-extra' or similar installed!"
    echo -n .; sleep 1; echo -n .; sleep 1; echo -n .; sleep 1; echo -n .; sleep 1; echo -n .; sleep 1; echo
fi

# PyWaveform
#have_pypkg waveform || test -f /usr/include/mpg123.h || \
#    { echo "ERROR: You MUST install mpg123 headers (e.g. 'apt-get install libmpg123-dev')!"; ok=false; }
have_pypkg waveform || test -f /usr/include/sndfile.h || \
    { echo "ERROR: You MUST install sndfile headers (e.g. 'apt-get install libsndfile1-dev')!"; ok=false; }

# numpy / scipy
have_pypkg scipy || which gfortran >/dev/null || \
    { echo "ERROR: You MUST install GNU fortran (e.g. 'apt-get install gfortran')!"; ok=false; }
have_pypkg numpy || test -d /usr/include/atlas || \
    { echo "ERROR: You MUST install ATLAS headers (e.g. 'apt-get install libatlas-dev')!"; ok=false; }
have_pypkg numpy || test -f /usr/lib/liblapack.a || \
    { echo "ERROR: You MUST install the LAPACK libraries (e.g. 'apt-get install liblapack-dev')!"; ok=false; }

$ok || { echo "FATAL: Fix above dependency errors first!"; return 1; }

# Tools
./bin/easy_install -U "setuptools>=0.6c11" || return 1
./bin/easy_install -U "paver>=1.0.1" || return 1
./bin/easy_install -U "nose>=1.0" || return 1
./bin/easy_install -U "yolk>=0.4.1" || return 1
./bin/easy_install -U "bpython" || return 1

# Dependencies (all optional, failures will lead to non-functional features)
fail=""
#have_pypkg waveform || easy_install -U https://github.com/pyroscope/PyWaveform/zipball/master || :
have_pypkg waveform || \
    ( git clone git://github.com/pyroscope/PyWaveform.git ; cd PyWaveform; ../bin/python setup.py install ) || \
      fail="$fail:PyWaveform install failure, creating waveform images won't work"
have_pypkg numpy || easy_install numpy || fail="$fail:numpy install failure, spectrograms won't work"
have_pypkg matplotlib || easy_install matplotlib || fail="$fail:matplotlib install failure, spectrograms won't work"
have_pypkg scipy || easy_install scipy || fail="$fail:scipy install failure, spectrograms won't work"

./bin/paver develop -U

test -z "$fail" || { echo "INSTALL FAILURES:"; tr : \\n <<<"$fail"; echo; }



src/auvyon/audio/transcode.py
""" Transcode audio formats.
"""
from __future__ import with_statement

import os
import tempfile
import subprocess
from contextlib import contextmanager

from auvyon import config


def extract_wav(datafile, target=None):
    """ Get LPCM 16bit audio stream from mediafile.
    
        If `target` is a directory, create a .wav with the same basename as the input.
        If target is empty, write to the directory of the source file. Otherwise, use it 
        directly as the target filename.
    """
    if datafile.endswith(".wav"):
        return datafile

    target = target or os.path.dirname(datafile) 
    if os.path.isdir(target):
        target = os.path.join(target, os.path.splitext(os.path.basename(datafile))[0] + ".wav")

    if datafile.endswith(".flac"):
        cmd = [config.CMD_FLAC, "--silent", "--decode", "--force", "-o", target, datafile]
    else:
        cmd = [config.CMD_FFMPEG, "-v", "0", "-y", "-i", datafile, "-acodec", "pcm_s16le", "-ac", "2", target]
    subprocess.check_call(cmd, stdout=open(os.devnull, "wb"), stderr=subprocess.STDOUT)

    return target


@contextmanager
def to_wav(mediafile):
    """ Context manager providing a temporary WAV file created from the given media file.
    """
    if mediafile.endswith(".wav"):
        yield mediafile
    else:
        wavfile = tempfile.mktemp(__name__) + ".wav"
        try:
            extract_wav(mediafile, wavfile)
            yield wavfile
        finally:
            if os.path.exists(wavfile):
                os.remove(wavfile)


src/auvyon/imaging/waveform.py
""" Audio data visualization.
"""
from __future__ import with_statement

import os
import sys
import shlex
import shutil
import tempfile
import subprocess

from auvyon import config
from auvyon.audio import transcode


# Default colors (RGBA)
WAVE_CENTER_COLOR = (0x00, 0xbb, 0x11, 0xff)
WAVE_OUTER_COLOR  = (0x99, 0x00, 0x11, 0xff)
WAVE_BG_COLOR     = (0x11, 0x11, 0x33, 0xff)
WAVE_LABEL_STYLE  = "-background '#333366ff' -stroke '#fff' -font Verdana-Standard"


def waveform_image(mediafile, xy_size, outdir=None, center_color=None, outer_color=None, bg_color=None):
    """ Create waveform image from audio data.
        Return path to created image file.
    """
    try:
        import waveform
    except ImportError, exc:
        raise ImportError("%s [get it at https://github.com/superjoe30/PyWaveform]" % exc)

    outdir = outdir or os.path.dirname(mediafile)
    outfile = os.path.join(outdir, os.path.splitext(os.path.basename(mediafile))[0] + ".png")

    with transcode.to_wav(mediafile) as wavfile:
        # Draw using a gradient
        waveform.draw(wavfile, outfile, xy_size, 
            bgColor=bg_color or WAVE_BG_COLOR,
            fgGradientCenter=center_color or WAVE_CENTER_COLOR, 
            fgGradientOuter=outer_color or WAVE_OUTER_COLOR)

    return outfile


def waveform_stack(mediafiles, xy_size, output=None, label_style=None, 
        center_color=None, outer_color=None, bg_color=None):
    """ Create a stack of waveform images from audio data.
        Return path to created image file.
    """
    img_files = []
    output = output or os.path.abspath(os.path.dirname(os.path.commonprefix(mediafiles)))
    if os.path.isdir(output):
        output = os.path.join(output, "waveforms.jpg")
    cmd = [config.CMD_IM_MONTAGE, "-tile", "1x%d" % len(mediafiles), "-geometry", "%dx%d" % xy_size, 
        "-label", "%t"] + shlex.split(label_style or WAVE_LABEL_STYLE)

    try:
        tempdir = tempfile.mktemp(__name__)
        os.makedirs(tempdir) 

        for mediafile in sorted(mediafiles):
            img_files.append(waveform_image(mediafile, xy_size, tempdir, center_color, outer_color, bg_color))

        cmd.extend(img_files)
        cmd.append(output)
        subprocess.check_call(cmd, stdout=open(os.devnull, "wb"), stderr=subprocess.STDOUT)
    finally:
        if os.path.isdir(tempdir):
            shutil.rmtree(tempdir, ignore_errors=True)

    return output


def _main():
    """ Command line interface for testing.
    """
    if len(sys.argv) <= 1:
        print("Usage: python -m auvyon.imaging.waveform <mediafile>...")
    else:
        try:
            print("Created %s" % waveform_stack(sys.argv[1:], (640, 48)))
        except subprocess.CalledProcessError, exc:
            print("Conversion error: %s" % exc)


if __name__ == "__main__":
    _main()


src/auvyon/imaging/waveforms.py
""" Audio data visualization.
"""
from __future__ import with_statement

import os
import sys
import shlex
import shutil
import tempfile
import subprocess

from auvyon import config
from auvyon.audio import transcode


# Default colors (RGBA)
WAVE_CENTER_COLOR = (0x00, 0xbb, 0x11, 0xff)
WAVE_OUTER_COLOR  = (0x99, 0x00, 0x11, 0xff)
WAVE_BG_COLOR     = (0x11, 0x11, 0x33, 0xff)
WAVE_LABEL_STYLE  = "-background '#333366ff' -stroke '#fff' -font DejaVu-Sans-ExtraLight -pointsize 12"


def waveform_image(mediafile, xy_size, outdir=None, center_color=None, outer_color=None, bg_color=None):
    """ Create waveform image from audio data.
        Return path to created image file.
    """
    try:
        import waveform
    except ImportError, exc:
        raise ImportError("%s [get it at https://github.com/superjoe30/PyWaveform]" % exc)

    outdir = outdir or os.path.dirname(mediafile)
    outfile = os.path.join(outdir, os.path.splitext(os.path.basename(mediafile))[0] + ".png")

    with transcode.to_wav(mediafile) as wavfile:
        # Draw using a gradient
        waveform.draw(wavfile, outfile, xy_size, 
            bgColor=bg_color or WAVE_BG_COLOR,
            fgGradientCenter=center_color or WAVE_CENTER_COLOR, 
            fgGradientOuter=outer_color or WAVE_OUTER_COLOR)

    return outfile


def waveform_stack(mediafiles, xy_size, output=None, label_style=None, 
        center_color=None, outer_color=None, bg_color=None):
    """ Create a stack of waveform images from audio data.
        Return path to created image file.
    """
    img_files = []
    output = output or os.path.abspath(os.path.dirname(os.path.commonprefix(mediafiles)))
    if os.path.isdir(output):
        output = os.path.join(output, "waveforms.jpg")
    cmd = [config.CMD_IM_MONTAGE] + shlex.split(label_style or WAVE_LABEL_STYLE)
    cmd += ["-tile", "1x%d" % len(mediafiles), "-geometry", "%dx%d" % xy_size, "-label", "%t"]

    try:
        tempdir = tempfile.mktemp(__name__)
        os.makedirs(tempdir) 

        for mediafile in sorted(mediafiles):
            img_files.append(waveform_image(mediafile, xy_size, tempdir, center_color, outer_color, bg_color))

        cmd.extend(img_files)
        cmd.append(output)
        subprocess.check_call(cmd, stdout=open(os.devnull, "wb"), stderr=subprocess.STDOUT)
    finally:
        if os.path.isdir(tempdir):
            shutil.rmtree(tempdir, ignore_errors=True)

    return output


def _main():
    """ Command line interface for testing.
    """
    if len(sys.argv) <= 1:
        print("Usage: python -m auvyon.imaging.waveforms <mediafile>...")
    else:
        try:
            print("Created %s" % waveform_stack(sys.argv[1:], (640, 48)))
        except subprocess.CalledProcessError, exc:
            print("Conversion error: %s" % exc)


if __name__ == "__main__":
    _main()


src/auvyon/imaging/spectrograms.py
""" Audio data sample spectrograms.
"""
from __future__ import with_statement

import os
import sys
import tempfile
import subprocess
from contextlib import closing

from auvyon import config
from auvyon.audio import transcode


def spectrogram_image(mediafile, dpi=72, outdir=None, outfile=None):
    # TODO: Add some of the constants below as parameters
    """ Create spectrogram image from audio data.
        Return path to created image file.
    """
    import matplotlib
    matplotlib.use('Agg')

    import matplotlib.pyplot as plt
    import scipy.io.wavfile
    import numpy as np
    import pylab

    # Output file path
    outfile = outfile or ""
    if outdir and outfile and os.sep in outfile:
        raise ValueError("Do not specify paths in both output directory '%s' and filename '%s'" % (outdir, outfile))

    if os.sep not in outfile:
        if not outfile:
            outfile = os.path.splitext(os.path.basename(mediafile))[0] + ".jpg"
        if not outdir:
            outdir = os.path.dirname(mediafile)
        outfile = os.path.join(outdir, outfile)

    with closing(open(os.devnull, "wb")) as black_hole:
        # Read audio data
        with transcode.to_wav(mediafile) as wavfile:
            sys.stdout, saved_stdout = black_hole, sys.stdout
            try:
                sample_rate, waveform = scipy.io.wavfile.read(wavfile)
            finally:
                sys.stdout = saved_stdout

        # Limit data to 10 second window from the middle, else the FFT needs ages
        data_window = sample_rate * 2 # secs
        waveform = [i[0] for i in waveform[(len(waveform) - data_window) // 2 : (len(waveform) + data_window) // 2]]
        # TODO: combine / add the channels to mono

        # Calculate FFT inputs
        nstep = int(sample_rate * 0.001) # 1ms step
        nfft = nwin = int(sample_rate * 0.005) & ~1 # 5ms window
        window = np.hamming(nwin)

        # Create spectrogram
        pylab.spectral()
        for khz in (5, 10, 16, 18, 20):
            pylab.text(data_window / sample_rate * .99, khz * 1000 + 75, "%d kHz" % khz, ha="right")
            pylab.axhline(khz * 1000)
        pylab.axis("off")
        pylab.specgram(waveform, NFFT=nfft, Fs=sample_rate, window=window)

        # Write to image
        try:
            pylab.savefig(outfile + ".png", format='png', facecolor="#000000", edgecolor="#000000", 
                dpi=dpi, transparent=True, bbox_inches="tight")

            cmd = [config.CMD_IM_CONVERT, "-trim", "-quality", "85", outfile + ".png", outfile]
            subprocess.check_call(cmd, stdout=black_hole, stderr=subprocess.STDOUT)
        finally:
            if os.path.exists(outfile + ".png"):
                os.remove(outfile + ".png")

    return outfile


def _main():
    """ Command line interface for testing.
    """
    if len(sys.argv) != 2:
        print("Usage: python -m auvyon.imaging.spectrograms <mediafile>")
    else:
        try:
            print("Created %s" % spectrogram_image(sys.argv[1], dpi=103, outfile="spectrogram.jpg"))
        except subprocess.CalledProcessError, exc:
            print("Conversion error: %s" % exc)


if __name__ == "__main__":
    _main()


pavement.py
# -*- coding: utf-8 -*-
""" AuVyon - Audio-Visual Tools.

    AuVyon collects little helpers to work with media data.

    Copyright (c) 2011 The PyroScope Project <pyroscope.project@gmail.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
"""
from __future__ import with_statement

import os
import re
import sys
import webbrowser

from paver.easy import *
from paver.setuputils import setup

from setuptools import find_packages


#
# Project Metadata
#

changelog = path("debian/changelog")

name, version = open(changelog).readline().split(" (", 1)
version, _ = version.split(")", 1)

project = dict(
    # egg
    name = "auvyon",
    version = version,
    package_dir = {"": "src"},
    packages = find_packages("src", exclude = ["tests"]),
    #entry_points = {
    #    "console_scripts": [
    #        "... = auvyon.scripts.xxx:run",
    #    ],
    #},
    include_package_data = True,
    zip_safe = True,
    data_files = [
        ("EGG-INFO", [
            "README", "LICENSE", "debian/changelog",
        ]),
    ],

    # dependencies
    install_requires = [
    ],
    setup_requires = [
    ],
    extras_require = {
        "waveforms": ["waveform>=0.4"],
    },

    # tests
    test_suite = "nose.collector",

    # cheeseshop
    author = "The PyroScope Project",
    author_email = "pyroscope.project@gmail.com",
    description = __doc__.split('.', 1)[0].strip(),
    long_description = __doc__.split('.', 1)[1].strip(),
    license = [line.strip() for line in __doc__.splitlines()
        if line.strip().startswith("Copyright")][0],
    url = "https://github.com/pyroscope/auvyon",
    keywords = "audio video metadata cli tool python",
    classifiers = [
        # see http://pypi.python.org/pypi?:action=list_classifiers
        "Development Status :: 4 - Beta",
        #"Development Status :: 5 - Production/Stable",
        "Environment :: Console",
        "Intended Audience :: End Users/Desktop",
        "Intended Audience :: Developers",
        "License :: OSI Approved :: GNU General Public License (GPL)",
        "Natural Language :: English",
        "Operating System :: OS Independent",
        "Programming Language :: Python :: 2.5",
        "Topic :: Multimedia",
        "Topic :: Multimedia :: Sound/Audio",
        "Topic :: Multimedia :: Video",
        "Topic :: Software Development :: Libraries :: Python Modules",
        "Topic :: Utilities",
    ],
)


#
# Helpers
#
def toplevel_packages():
    """ Get package list, without sub-packages.
    """
    packages  = set(options.setup.packages)
    for pkg in list(packages):
        packages -= set(p for p in packages if str(p).startswith(pkg + '.'))
    return list(sorted(packages))


#
# Build
#
@task
@needs(["setuptools.command.egg_info"])
def bootstrap():
    """ Initialize project.
    """


@task
@cmdopts([
    ('docs-dir=', 'd', 'directory to put the api documentation in'),
    ('excludes=', 'x', 'list of packages to exclude'),
])
def docs():
    """ Create documentation.
    """
    from epydoc import cli

    path('build').exists() or path('build').makedirs()

    # get storage path
    docs_dir = options.docs.get('docs_dir', 'docs/apidocs')

    # clean up previous docs
    (path(docs_dir) / "epydoc.css").exists() and path(docs_dir).rmtree()

    # set up excludes
    try:
        exclude_names = options.docs.excludes
    except AttributeError:
        exclude_names = []
    else:
        exclude_names = exclude_names.replace(',', ' ').split()

    excludes = []
    for pkg in exclude_names:
        excludes.append("--exclude")
        excludes.append('^' + re.escape(pkg))

    # call epydoc in-process
    sys_argv = sys.argv
    try:
        sys.argv = [
            sys.argv[0] + "::epydoc",
            "-v",
            "--inheritance", "listed",
            "--output", docs_dir,
            "--name", "%s %s" % (options.setup.name, options.setup.version),
            "--url", options.setup.url,
            "--graph", "umlclasstree",
        ] + excludes + toplevel_packages()
        sys.stderr.write("Running '%s'\n" % ("' '".join(sys.argv)))
        cli.cli()
    finally:
        sys.argv = sys_argv


@task
@needs("docs")
def dist_docs():
    """ Create a documentation bundle.
    """
    dist_dir = path("dist")
    docs_package = path("%s/%s-%s-docs.zip" % (dist_dir.abspath(), options.setup.name, options.setup.version))

    dist_dir.exists() or dist_dir.makedirs()
    docs_package.exists() and docs_package.remove()

    sh(r'cd docs && find . -type f \! \( -path "*/.svn*" -o -name "*~" \) | sort'
       ' | zip -qr -@ %s' % (docs_package,))

    print
    print "Upload @ http://pypi.python.org/pypi?:action=pkg_edit&name=%s" % ( options.setup.name,)
    print docs_package


#
# Testing
#

@task
@needs("nosetests")
def test():
    """ Run unit tests.
    """


@task
def coverage():
    """ Generate coverage report and show in browser.
    """
    coverage_index = path("build/coverage/index.html")
    coverage_index.remove()
    sh("paver test")
    coverage_index.exists() and webbrowser.open(coverage_index)


@task
@needs("setuptools.command.build")
def functest():
    """ Functional test of the command line tools.
    """
    sh("echo nothing at the moment")


#
# Release Management
#
@task
@needs("clean")
def dist_clean():
    """ Clean up including dist directory.
    """
    path("dist").rmtree()


@task
@needs(["dist_clean", "minilib", "generate_setup", "sdist"])
def release():
    """ Check release before upload to PyPI.
    """
    sh("paver bdist_egg")

    # Check that source distribution can be built and is complete
    print
    print "~" * 78
    print "TESTING SOURCE BUILD"
    sh(
        "{ cd dist/ && unzip -q %s-%s.zip && cd %s-%s/"
        "  && /usr/bin/python setup.py sdist >/dev/null"
        "  && if { unzip -ql ../%s-%s.zip; unzip -ql dist/%s-%s.zip; }"
        "        | cut -b26- | sort | uniq -c| egrep -v '^ +2 +' ; then"
        "       echo '^^^ Difference in file lists! ^^^'; false;"
        "    else true; fi; } 2>&1"
        % tuple([project["name"], version] * 4)
    )
    path("dist/%s-%s" % (project["name"], version)).rmtree()
    print "~" * 78

    print
    print "Created", " ".join([str(i) for i in path("dist").listdir()])
    print "Use 'paver sdist bdist_egg upload' to upload to PyPI"
    print "Use 'paver dist_docs' to prepare an API documentation upload"


#
# Other
#
@task
@cmdopts([
    ('output=', 'o', 'Create report file (.html, .log, or .txt) [stdout]'),
    ('rcfile=', 'r', 'Configuration file [./pylint.cfg]'),
    ('msg-only', 'm', 'Only generate messages (no reports)'),
])
def lint():
    """ Report pylint results.
    """
    from pylint import lint as linter

    # report according to file extension
    reporters = {
        ".html": linter.HTMLReporter,
        ".log": linter.ParseableTextReporter,
        ".txt": linter.TextReporter,
    }

    lint_build_dir = path("build/lint")
    lint_build_dir.exists() or lint_build_dir.makedirs()

    argv = []
    rcfile = options.lint.get("rcfile")
    if not rcfile and path("pylint.cfg").exists():
        rcfile = "pylint.cfg"
    if rcfile:
        argv += ["--rcfile", os.path.abspath(rcfile)]
    if options.lint.get("msg_only", False):
        argv += ["-rn"]
    argv += [
        "--import-graph", (lint_build_dir / "imports.dot").abspath(),
    ]
    argv += toplevel_packages()

    sys.stderr.write("Running %s::pylint '%s'\n" % (sys.argv[0], "' '".join(argv)))
    outfile = options.lint.get("output", None)
    if outfile:
        outfile = os.path.abspath(outfile)

    try:
        with pushd("src" if path("src").exists() else "."):
            if outfile:
                reporterClass = reporters.get(path(outfile).ext, linter.TextReporter)
                sys.stderr.write("Writing output to %r\n" % (str(outfile),))
                linter.Run(argv, reporter=reporterClass(open(outfile, "w")))
            else:
                linter.Run(argv)
    except SystemExit, exc:
        if not exc.code:
            sys.stderr.write("paver::lint - No problems found.\n")
        elif exc.code & 32:
            # usage error (internal error in this code)
            sys.stderr.write("paver::lint - Usage error, bad arguments %r?!\n" % (argv,))
            raise
        else:
            bits = {
                1: "fatal",
                2: "error",
                4: "warning",
                8: "refactor",
                16: "convention",
            }
            sys.stderr.write("paver::lint - Some %s message(s) issued.\n" % (
                ", ".join([text for bit, text in bits.items() if exc.code & bit])
            ))
            if exc.code & 3:
                sys.stderr.write("paver::lint - Exiting due to fatal / error message.\n")
                raise


#
# Main
#
setup(**project)
